fix: use subquery with join map one methods

xdc0 opened the pull request:
<!--
  😀 Wonderful!  Thank you for opening a pull request for TypeORM.

  Please fill in the information below to expedite the review
  and (hopefully) merge of your change.

  If unsure about something.. just do as best as you're able,
  or reach out through our community support channels!
  https://github.com/typeorm/typeorm/blob/master/docs/support.md
-->

### Description of change

On https://github.com/typeorm/typeorm/pull/9354 there was a patch to map subquery results to an entity when using the `leftJoinAndMapOne` and `innerJoinAndMapOne` methods through a new parameter: `mapAsEntity`.

This unfortunately does not quite work as expected because the subquery is ignored since the join logic detects that there is metadata and the logic will override the subquery in favor of generated queries via metadata.

This problem is described by @cjsauer on the following comment: https://github.com/cjsauer

### Pull-Request Checklist

<!--
  Please make sure to review and check all of the following.

  If an item is not applicable, you can add "N/A" to the end.
-->

- [x] Code is up-to-date with the `master` branch
- [x] `npm run format` to apply prettier formatting
- [x] `npm run test` passes with this change
- [x] This pull request links relevant issues as `Fixes #0000`
- [x] There are new or updated unit tests validating the change
~- [ ] Documentation has been updated to reflect this change~
- [x] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

<!--
  🎉 Thank you for contributing and making TypeORM even better!
-->

xdc0 commented on src/query-builder/SelectQueryBuilder.ts:
```diff
@@ -2063,20 +2084,6 @@ export class SelectQueryBuilder<Entity extends ObjectLiteral>
                 })
             }
         } else {
-            let subQuery: string = ""
-            if (typeof entityOrProperty === "function") {
-                const subQueryBuilder: SelectQueryBuilder<any> = (
-                    entityOrProperty as any
-                )((this as any as SelectQueryBuilder<any>).subQuery())
-                this.setParameters(subQueryBuilder.getParameters())
-                subQuery = subQueryBuilder.getQuery()
-            } else {
-                subQuery = entityOrProperty
-            }
-            const isSubQuery =
-                typeof entityOrProperty === "function" ||
-                (entityOrProperty.substr(0, 1) === "(" &&
-                    entityOrProperty.substr(-1) === ")")
```
Pulled this up above the `if/else` block since it's used on both conditions now.

xdc0 commented on src/query-builder/SelectQueryBuilder.ts:
```diff
@@ -2034,6 +2034,26 @@ export class SelectQueryBuilder<Entity extends ObjectLiteral>
         joinAttribute.condition = condition // joinInverseSideCondition
         // joinAttribute.junctionAlias = joinAttribute.relation.isOwning ? parentAlias + "_" + destinationTableAlias : destinationTableAlias + "_" + parentAlias;
         this.expressionMap.joinAttributes.push(joinAttribute)
+        const isEntity = this.connection.hasMetadata(entityOrProperty)
+        const isSubQuery =
+            (!isEntity && typeof entityOrProperty === "function") ||
```
Class references will give `function` when calling `typeof foo`. Now the argument is determined to be an entity if the connection has metadata.

xdc0 commented on src/query-builder/SelectQueryBuilder.ts:
```diff
@@ -2051,6 +2071,7 @@ export class SelectQueryBuilder<Entity extends ObjectLiteral>
                 type: "join",
                 name: aliasName,
                 metadata: joinAttributeMetadata,
+                subQuery: isSubQuery ? subQuery : undefined,
```
When subquery is provided, https://github.com/typeorm/typeorm/blob/58c243d07c319a4eb4b218889665de0d04c8f4ff/src/query-builder/SelectQueryBuilder.ts#L2276-L2278 `createJoinExpression` already prefers using a subQuery if present, it doesn't seem to have any other effect besides this.

If there is any potential problem of passing `subQuery` here, please chime in.

0xtrou commented:
Need this to get merged asap

pzcuong commented:
Need this to get merged asap

narimanam commented:
any hope on this merge?

gioboa reviewed (commented):
Thanks @xdc0 
is this PR still valid? Could you add tests to cover this fix?

xdc0 commented:
@gioboa It's still be valid. The PR includes test coverage here: https://github.com/typeorm/typeorm/pull/10769/files#diff-fc1093be2aa4e9334350e29375146e0e2cc029295285b80c5caf92dac490c422

Open to comments if it's in need of more tests.

xdc0 commented on test/functional/query-builder/join/query-builder-joins.ts:
```diff
@@ -570,6 +570,40 @@ describe("query builder > joins", () => {
+                        .leftJoinAndMapOne(
+                            "post.tag",
+                            (qb) =>
+                                qb
```
Updated test to use `subQuery().from(...)`

Note: Other tests in this suite do not use `subQuery()`, I'll leave updating unrelated tests an exercise for another time.

xdc0 commented:
Updated fork with upstream ✅ 
Updated test to use `subQuery` and verified issue still persists after that change ✅ 

Ready for another review pass.

xdc0 commented:
> ## PR Code Suggestions ✨
> **Category**	**Suggestion                                                                                                                                    **	**Impact**
> General	
> Replace deprecated substr method
> **Replace the deprecated `substr()` method with `charAt()` for better code maintainability and future compatibility.**
> 
> [src/query-builder/SelectQueryBuilder.ts [2099-2104]](https://github.com/typeorm/typeorm/pull/10769/files#diff-52bea788267888fdbb5f795f0dfb61504f16bd43e0afdfa28848abb4435121b5R2099-R2104)
> 
> ```diff
>  const isEntity = this.connection.hasMetadata(entityOrProperty)
>  const isSubQuery =
>      (!isEntity && typeof entityOrProperty === "function") ||
>      (typeof entityOrProperty === "string" &&
> -        entityOrProperty.substr(0, 1) === "(" &&
> -        entityOrProperty.substr(-1) === ")")
> +        entityOrProperty.charAt(0) === "(" &&
> +        entityOrProperty.charAt(entityOrProperty.length - 1) === ")")
> ```
> 
> * [ ]  **Apply / Chat**
> 
> Suggestion importance[1-10]: 4
> __
> 
> Why: The suggestion correctly identifies the use of the deprecated `substr()` method and proposes a valid replacement with `charAt()`, which improves code maintainability and aligns with modern JavaScript standards.
> 
> Low
> * [ ]  More

This is not new code added in this PR. I'm merely moving up that evaluation without touching any of the current implementation

pkuczynski commented on src/query-builder/SelectQueryBuilder.ts:
```diff
@@ -2096,6 +2096,26 @@ export class SelectQueryBuilder<Entity extends ObjectLiteral>
+                this.setParameters(subQueryBuilder.getParameters())
+                subQuery = subQueryBuilder.getQuery()
+            }
+        }
```
Simpler, more readable, more modern js

pkuczynski commented on src/query-builder/SelectQueryBuilder.ts:
```diff
@@ -2113,6 +2133,7 @@ export class SelectQueryBuilder<Entity extends ObjectLiteral>
                 type: "join",
                 name: aliasName,
                 metadata: joinAttributeMetadata,
+                subQuery: isSubQuery ? subQuery : undefined,
```
```suggestion
                subQuery,
```

gioboa reviewed (approved):
LGTM, thank you @xdc0
